modules/mauth-authenticator-http4s/src/main/scala/com/mdsol/mauth/http4s/RequestAuthenticator.scala

```diff
@@ -0,0 +1,155 @@
+      val message = "The service requires mAuth v2 authentication headers."
+      logger.error(message) *> Async[F].raiseError(new MAuthValidationException(message))
+    } else {
+      getPublicKey(mAuthRequest)
+    }
+  }
+
+  private def getPublicKey(mAuthRequest: MAuthRequest): F[Boolean] = {
+    publicKeyProvider.getPublicKey(mAuthRequest.getAppUUID).map {
+      case None =>
+        logger.error("Public Key couldn't be retrieved")
+        false
+      case Some(clientPublicKey) =>
+        // Decrypt the signature with public key from requesting application.
+        mAuthRequest.getMauthVersion match {
+          case MAuthVersion.MWS =>
+            logger.warn("MAuth v1 client was used to authenticate this request which is deprecated")
+            validateSignatureV1(mAuthRequest, clientPublicKey)
+          case MAuthVersion.MWSV2 =>
+            val v2IsValidated = validateSignatureV2(mAuthRequest, clientPublicKey)
+            if (isV2OnlyAuthenticate)
+              v2IsValidated
+            else if (v2IsValidated)
+              v2IsValidated
+            else
+              fallbackValidateSignatureV1(mAuthRequest, clientPublicKey)
+        }
+    }
+  }
+
+}
+
+object RequestAuthenticator {
+  def apply[F[_]: Async](publicKeyProvider: ClientPublicKeyProvider[F], epochTimeProvider: EpochTimeProvider) =
+    new RequestAuthenticator(publicKeyProvider, epochTimeProvider, v2OnlyAuthenticate = false)
+
+  // Check epoch time is not older than specified interval.
+  private def validateTime(requestTime: Long, epochTimeProvider: EpochTimeProvider)(requestValidationTimeout: Duration): Boolean =
+    (epochTimeProvider.inSeconds - requestTime) < requestValidationTimeout.toSeconds
+
+  // Check V2 header if only V2 is required
+  private def validateMauthVersion(mAuthRequest: MAuthRequest, v2OnlyAuthenticate: Boolean): Boolean =
+    !v2OnlyAuthenticate || mAuthRequest.getMauthVersion == MAuthVersion.MWSV2
+
+  // check signature for V1
+  private def validateSignatureV1[F[_]](mAuthRequest: MAuthRequest, clientPublicKey: PublicKey)(implicit logger: SelfAwareStructuredLogger[F]): Boolean = {
+    logAuthenticationRequest(mAuthRequest)
+    val decryptedSignature = MAuthSignatureHelper.decryptSignature(clientPublicKey, mAuthRequest.getRequestSignature)
+    // Recreate the plain text signature, based on the incoming request parameters, and hash it.
+    try {
+      val messageDigest_bytes = MAuthSignatureHelper.generateDigestedMessageV1(mAuthRequest).getBytes(StandardCharsets.UTF_8)
+
+      // Compare the decrypted signature and the recreated signature hashes.
+      // If both match, the request was signed by the requesting application and is valid.
+      util.Arrays.equals(messageDigest_bytes, decryptedSignature)
+    } catch {
+      case ex: Exception =>
+        val message = "MAuth request validation failed for V1."
+        logger.error(ex)(message)
+        throw new MAuthValidationException(message, ex)
+    }
```

Comment:
Nope - you can definitely do better than try/catch and rethrow (we're in cats effect land, you can mark failed computations as failed `F`, so that the compiler has full control over the correctness of your program). Also, mind the chaining of the logger statements, otherwise they won't be executed